src/watch_filter.rs

```diff
@@ -0,0 +1,313 @@
+    pub(crate) fn new(
+        app_dir: PathBuf,
+        path_patterns: Vec<String>,
+        ignore_patterns: Vec<String>,
+    ) -> Self {
+        // Record approximate start time of process for the purpose of making sure files have been modified since
+        let start_time = SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .unwrap()
+            .as_micros();
+
+        tracing::info!("watching relative to app dir: {:?}", app_dir);
+        tracing::info!("watching path patterns: {:?}", path_patterns);
+        tracing::info!("ignoring path patterns: {:?}", ignore_patterns);
+
+        WatchFilter {
+            path_patterns: path_patterns
+                .iter()
+                .map(|path| Pattern::new(app_dir.join(path).to_str().unwrap()).unwrap())
+                .collect(),
+            ignore_patterns: ignore_patterns
+                .iter()
+                .map(|path| Pattern::new(app_dir.join(path).to_str().unwrap()).unwrap())
+                .collect(),
+            process_start_since_epoch: start_time,
+        }
+    }
+
+    fn has_valid_event_kind(&self, event: &Event) -> bool {
+        event.tags.iter().any(|tag| {
+            matches!(
+                tag,
+                Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_)))
+                    | Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Name(_)))
+                    | Tag::FileEventKind(FileEventKind::Create(_))
+                    | Tag::FileEventKind(FileEventKind::Remove(_))
+            )
+        })
+    }
+
+    fn matches_one_of_ignore_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.ignore_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    fn matches_one_of_path_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.path_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    /// This is necessary to check due to a bug on macOS emitting modify events on copies
+    /// https://github.com/rust-lang/rust/issues/107130
+    fn path_has_been_actually_modified(&self, event: &Event) -> Result<bool> {
+        // No need to check if a deleted path has been modified because it is gone
+        if event
+            .tags
+            .iter()
+            .any(|tag| matches!(tag, Tag::FileEventKind(FileEventKind::Remove(_))))
+        {
+            return Ok(true);
+        }
+
+        for (path, _) in event.paths() {
+            let metadata = std::fs::metadata(path)?;
+            let time_since_epoch = metadata
+                .modified()?
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_micros();
+            let mut modified_map = FILES_MODIFIED_AT.lock().unwrap();
+            let path_key = path.to_str().unwrap().to_owned();
```

Comment:
Probably shouldn't `unwrap` here. It isn't super common top have non-unicode file names, but its also pretty nasty to panic if it happens. Maybe just log and `continue` on error?